### ISSUE TITLE
High memory requirements during feature extraction

### ISSUE THREAD

**Issue by hoesler** [blue-yonder/tsfresh#702]
1. Your operating system
Linux

2. The version of *tsfresh* that you are using
0.16.0

3. The data on which the problem occurred (please do not upload 1000s of time series but try to boil the problem down to a small group or even a singular one)
```
<class 'pandas.core.frame.DataFrame'>
Int64Index: 6024604 entries, 0 to 1299438
Columns: 77 entries, index to machine_id
dtypes: float32(75), int64(1), object(1)
memory usage: 1.8+ GB
```

4. A minimal code snippet which reproduces the problem/bug
See below

5. Any reported errors or traceback
OOM

I was experimenting with tsfresh on the above mentioned dataset using a scikit learn pipeline:
`RelevantFeatureAugmenter(column_id='machine_id', column_sort='index', default_fc_parameters = MinimalFCParameters(), n_jobs=4, chunksize=10)
`
The code was executed from a Jupyter notebook which has 20GB of RAM available. The kernel died from out of memory errors, no matter which parameters I chose for `chunksize` or `n_jobs`. This seemed a bit strange cosidering the medium sized input and the tasks I was imagining tsfresh to do. My first idea was to fit (select features) only on a sample of the train data. This worked well, but the feature extraction during the transform step of the ~70 relevant features was still causing the same problem.
The next idea was scaling out. But before trying to distribute the work using dask, I started to trace at which step the memory starts to inflate and found that it the problem is the [sorting](https://github.com/blue-yonder/tsfresh/blob/a0039838a8264cbe8fb9f512849b778dfc3ccc8c/tsfresh/utilities/dataframe_functions.py#L359) of the [melted](https://github.com/blue-yonder/tsfresh/blob/a0039838a8264cbe8fb9f512849b778dfc3ccc8c/tsfresh/utilities/dataframe_functions.py#L338) dataframe while transforming the data into the normalized format. This doesn't seem too strange if you consider that it is actually sorting 6.024.604 * 75 = 451.845.300 rows.

Because this data transformation is done before sending any data to the distributor, scaling out won't be a solution here.

So I forked tsfresh and started to shift the heavy transformations from the "driver" to the workers and ended with skipping the melt transformation completely. A more straight forward approach was to simply iterate the wide format.

You can find my first draft here:
https://github.com/blue-yonder/tsfresh/compare/v0.16.0...hoesler:distribute-melt?expand=1

Using my modified code, the kernel barely exceeded 8GB during the fit step on the complete train data set. It seemed a bit slower, though I didn't do any benchmarks. But that is totally acceptable for me, especially because this could be easy improved by scaling out.

I would love to finalize that into a pull request, but I'd like to hear your opinion first.

Cheers
Christoph

**Comment by nils-braun** [comment]
Hi @hoesler!
Thank you very much for your work and your observations.
Much appreciated!
As far as I understand your work, you are moving the data transformation from the normalisation to the chunk creation. I think that makes sense and as all of these are internal interfaces, that should not break anything else.

The normalisation is actually something we wanted to get rid for a long time (but it turned out to be very useful in the first iteration of the package). Thank you for working on this.

If you want, please continue with your work. I guess we can benchmark it in the end.
If you need any help or support, please tell me!

**Comment by adalseno** [comment]
Hi, I have a problem of high memory usage.
Mac OS X 10.14.6
tsfresh version 0.15.1 (conda) with python 3.7.7 on jupyter notebook
Dataset with 4.7 million rows and 14 columns (size around 0.5 GB)
I made feature extraction and selection on a small subsample of the dataset and identified 57 features. I saved them into a dict, `selected_features`, and made a full feature extraction on the whole df with:
`extract_features(df,
    column_id='id',
    column_sort='date', 
    default_fc_parameters=selected_features)`
Im my mind `tsfresh` should calculate only the (57) selected features, but the process took too long and the df grew in size over 70GB then I had to stop it. What can I do to keep RAM usage under control and, possibly, speed up the process?
Thanks.
